### PR TITLE
guide: Basic Operations (Data Mgmt)

### LINKED ISSUE #2856: guide: Data Management
UPDATE: https://github.com/iterative/dvc.org/issues/2856#issuecomment-1278366397

---

This is the plan for data management trail that focuses on:

- Adding data to DVC projects
  & Versioning data in DVC projects
- The cache (local)
- & [Shared cache](https://dvc.org/doc/user-guide/how-to/share-a-dvc-cache) (external)
- Removing data from DVC projects
- \+ Creating remotes (link to config/remotes)
- & Sync with remotes
  See also https://github.com/iterative/dvc.org/issues/2866
- Accessing public datasets and data registries (get, import)
- External data topics (See https://github.com/iterative/dvc.org/issues/563#issuecomment-1198687900)
  + probably address https://github.com/iterative/dvc.org/issues/520

<details>

## Adding data to DVC projects

- Initialize a DVC repository and use `dvc add` to add files. 

- We'll assume MNIST data exist in a folder and will add it. 
 
## Versioning data in DVC projects

- Overwrite Fashion-MNIST data on top of MNIST and update the dataset.
- Go back and forth in Git history to get different datasets in the same folder.

## Creating remotes

- Add a Google Drive folder as a remote.

- Make it default

## Pushing to/pulling from remotes

- Push the cache to the remote we created
- Clone the repository to somewhere (e.g. ssh or local folder) 
- Pull the cache
 
## Accessing public datasets and registries

- Get the Fashion MNIST data from dataset-registry

## Removing data from DVC projects

- Remove certain folders from workspace
- Delete the corresponding cache files

</details>

UPDATE: start with a reorg, see https://github.com/iterative/dvc.org/issues/2856#issuecomment-1110477162 below (may be enough).

### PR DESCRIPTION
**Pull request by jorgeorpinel**
To finish addressing the 2nd check box in https://github.com/iterative/dvc.org/issues/2856#issuecomment-1278366397. Planned structure:

- Tracking
- Synchronizing
  - Other forms of access? (`list`, gets, imports & `update`)
- Versioning (unifying aspect)

Main file to review: [UG/data-management/track-sync-version.md](https://github.com/iterative/dvc.org/pull/4053/files#diff-b9a35a5e5d2cd037cb369f3cf7c0865829b330c8d67653501dad18ec687f50db)

In review app: https://dvc-org-guide-data-mgmt-epdkkq.herokuapp.com/doc/user-guide/data-management/track-sync-data

UPDATE: Above done ✅ 

- [ ] Rewrite with more diagrams and workflow recipe focus? See https://github.com/iterative/dvc.org/pull/4053#issuecomment-1327917535 for more details.

---

- [ ] Do not merge before https://github.com/iterative/dvc.org/pull/4042

---

- [ ] Along with #4062, this should close #2856.

### CODE REVIEW

File: content/docs/user-guide/data-management/track-sync-version.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+workspace and updates `.dvc` files accordingly (changes visible to Git). If you
+need to move or rename tracked data without content changes, use `dvc move`.
+
+Finally, to stop tracking data, use `dvc remove`. To also remove it from the
+cache (either latest or historic versions), use `dvc gc`. See [more details].
+
+<admon type="tip">
+
+Other commands related to tracking data: `dvc unprotect`, `dvc import`, and
+`dvc import-url`.
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/track-sync-version.md:44]: I wonder if we should also mention (and link to) the `Repo` "API" here.

Comment (skshetry) [reply content/docs/user-guide/data-management/track-sync-version.md:45]: Please don't mention `Repo` API anywhere, unless it is something similar to https://www.mercurial-scm.org/wiki/MercurialApi.

File: content/docs/command-reference/data/status.md [hunk 2]
```diff
@@ -36,7 +36,7 @@ you don't accidentally commit something you don't mean to.
 
 An example output might look something like follows:
 
-```dvc
+```cli
```
Comment (jorgeorpinel) [inline content/docs/command-reference/data/status.md:39]: > Unrelated, sorry (I copied that block to the new file in this PR so decided to fix the highlighter in both places).

File: content/docs/user-guide/data-management/index.md [hunk 3]
```diff
@@ -66,14 +66,20 @@ data cache.
...
+branching, pull requests, etc.). This way machine learning teams can apply
+mature software engineering practices.
 
 ![Versioned ML project](/img/versioned-project.png) _Navigate versions with Git
 commits_
 
+The resulting projects are neatly organized in the "space dimension", having
+only the files and directories needed at the time and without complicated, ad
+hoc file names like `2022-10-20_linear-model_v2-Carl`. Project versions live in
+the "time dimension" ([Git history]).
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/index.md:81]: Updated general data versioning info in parent page (DM index)

File: content/docs/user-guide/data-management/track-sync-version.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+You may have noticed that most of the tracking and synchronization commands give
+out hints about `git` commands to follow DVC operations. That's because the
+unifying aspect across DVC features (for data management and beyond) is [data
+versioning].
+
+The way this looks is that many DVC operations write small [metafiles] to the
+<abbr>workspace</abbr>, which you can in turn track and version with Git, a
+feature-rich and battle-tested [SCM] tool.
+
+![Versioning flow](/img/flow.png) _Data versioning flow on top of Git_
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/track-sync-version.md:128]: Found this image unused already in the repo! Pretty close to what we need here:

image

> p.s. we should probably prune unused assets from [static/](https://github.com/iterative/dvc.org/tree/main/static) ?

File: content/docs/command-reference/remote/index.md [hunk 5]
```diff
@@ -24,37 +24,45 @@ positional arguments:
...
+`[azure]`, `[gdrive]`, `[gs]`, `[oss]`, `[ssh]`. Use `[all]` to include them
+all. For example:
+
+```cli
+$ pip install "dvc[s3]"
+```
+
+> This installs the `boto3` library along with DVC to support S3 storage.
+
+[installed dvc]: /doc/install
```
Comment (jorgeorpinel) [inline content/docs/command-reference/remote/index.md:58]: - [ ] Apply this change to the other docs that use this ref (separate PR).

File: content/docs/command-reference/remote/index.md [hunk 6]
```diff
@@ -24,49 +24,57 @@ positional arguments:
...
+<admon icon="book">
 
-For an intro on DVC remote usage see [Storing and sharing data].
+For an intro on DVC remote usage see [Storing and sharing data]. See also
+[Synchronizing data] for further general info about remote storage usage.
 
 [storing and sharing data]:
   /doc/start/data-management/data-versioning#storing-and-sharing
+[synchronizing data]:
+  /doc/user-guide/data-management/track-sync-version#synchronizing-data
```
Comment (jorgeorpinel) [inline content/docs/command-reference/remote/index.md:73]: Idk if we need both links here.

File: content/docs/user-guide/data-management/track-sync-version.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
+# Track, Sync, and Version Data
```
Comment (shcheklein) [inline content/docs/user-guide/data-management/track-sync-version.md:1]: Good stuff, @jorgeorpinel . But let's please step back and discuss the bigger picture first. How is it (expect being more low level) from all the get started, tutorials, etc? What would be the purpose of this page

Comment (jorgeorpinel) [reply content/docs/user-guide/data-management/track-sync-version.md:1]: Sounds good. I just wanted to get moving with all of this and ended up writing very final-looking docs (around one PR per day), but I realize we may need to reconsider a lot of the content, the approach, conceptual level, etc. In that sense these are still draft-ish. And thanks

File: content/docs/user-guide/data-management/track-sync-data.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+<admon type="tip">
+
+The `dvc commit` and `dvc checkout` plumbing commands let you force-sync them.
+This can be useful when unexpected errors occur (e.g. cache corruption).
+
+</admon>
+
+To enable additional storage locations, you can configure [DVC remotes] with the
+`dvc remote add` and `dvc remote modify` commands (see `dvc remote` for more
+options). Once this is done, use the `dvc push` and `dvc pull` commands to
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/track-sync-data.md:109]: > you can configure [DVC remotes] with the `dvc remote add` and `dvc remote modify` commands (see `dvc remote` for more options).

- [ ] After https://github.com/iterative/dvc.org/pull/4058, replace for link to /doc/user-guide/data-management/remote-storage

File: content/docs/user-guide/data-management/track-sync-version.md (hunk 7 above)
Comment (jorgeorpinel) [reply content/docs/user-guide/data-management/track-sync-version.md]: > How is it (different) from all the get started, tutorials, etc? What would be the purpose of this page?

This page is more comprehensive than the GS. It's not a tutorial because we'll keep this up to date regularly, and doesn't use a real-life example

- GS: high-level; strategic features only; Purpose: onboard
- UG: mixed; comprehensive (feature set scope); **Purpose: understand**
- CR: low-level; comprehensive (command scope); Purpose: usage

> Refer to our current [docs framework](https://www.notion.so/iterative/Framework-7e86e484e50e42529df3dbfe32b4e987) proposal.

Comment (jorgeorpinel) [reply content/docs/user-guide/data-management/track-sync-version.md]: > p.s. It may be more of a priority to review https://github.com/iterative/dvc.org/pull/4058 though, @shcheklein (after https://github.com/iterative/dvc.org/issues/2856).

**Comment by shcheklein** [comment]
@jorgeorpinel qq - is it something that waits on my feedback? (sorry lost in the comments a bit, just to make sure we are not blocked by me)

**Comment by jorgeorpinel** [comment]
@shcheklein sorry I, started many PRs (trying to move fast) so it's easy to get lost. But I only direclty requested a review in #4042 so far.

> Although, really I want to get to #4058 (which it's nested into #4042). Those 2 are marked p1.

**Comment by jorgeorpinel** [comment]
p.s. for this one @dberenbaum or others from the core team may be able to review, but they seem pretty busy ATM and there's no huge rush to get this in.

**Comment by dberenbaum** [comment]
@jorgeorpinel Do you want me to review #4058 or #4042?

**Comment by jorgeorpinel** [comment]
Simplify a lot? Avoid repeating content from GS and CR -- maybe add a condensed table of all the commands (index), focus on condensed diagrams. More valuable: go over workflows or scenarios.

> p.s. apply to https://dvc.org/doc/user-guide/pipelines/defining-pipelines so we can move fwd faster with that section as well. Ideas: recipes (e.g. from notebook to stages). Avoid formal descriptions (already in refs).

File: content/docs/user-guide/data-management/track-sync-data.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+`.dvc` files can be tracked (and [versioned](#versioning-data)) with Git.
+
+</admon>
+
+If you need to move or rename tracked data, use `dvc move`. To stop tracking
+data, use `dvc remove`. To also remove it from the cache, use `dvc gc`. See
+[more details].
+
+Putting it all together, we can get an overview of the data in a project with
+`dvc data status`. This will list changes to DVC-tracked data as well as files
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/track-sync-data.md:32]: Not sure why `dvc data status` is not auto-linking (see [in deployment](https://dvc-org-guide-data-mgmt-epdkkq.herokuapp.com/doc/user-guide/data-management/track-sync-data#tracking-data)). I see it in https://github.com/iterative/gatsby-theme-iterative/blob/main/packages/gatsby-theme-iterative/config/prismjs/dvc-commands.js . Cc @iterative/websites

**Comment by jorgeorpinel** [comment]
Alright. So [this one](https://dvc-org-guide-data-mgmt-epdkkq.herokuapp.com/doc/user-guide/data-management/track-sync-data) is done as originally intended (see OP).

Questions:

- Is it too long? May not be useful in its current form. See https://github.com/iterative/dvc.org/pull/4053#issuecomment-1327917535 for ideas on how to rewrite it.
- Do we need the [Versioning section](https://dvc-org-guide-data-mgmt-epdkkq.herokuapp.com/doc/user-guide/data-management/track-sync-data#versioning-data) given that there will be a Versioning page per https://github.com/iterative/dvc.org/pull/4119#pullrequestreview-1227376683 ?

I'll leave it up to you @dberenbaum @shcheklein . Turned it into a draft for now. Thanks

**Comment by jorgeorpinel** [comment]
> p.s. the check that's failing is on an unrelated file (content/docs/studio/user-guide/projects-and-experiments/live-metrics-and-plots.md). Seems like it was merged to `main` without proper format cc @iterative/websites if you want to take a look. Thanks

**Comment by shcheklein** [comment]
> p.s. the check that's failing is on an unrelated file (content/docs/studio/user-guide/projects-and-experiments/live-metrics-and-plots.md

@jorgeorpinel  Hmm, I can't reproduce locally (`yarn format-all` doesn't return anything). Could you please rebase / merge main into the target branch?

**Comment by jorgeorpinel** [comment]
> yarn format-all doesn't return anything

Its the same for me. Yet [the check fails](https://github.com/iterative/dvc.org/actions/runs/4219476407/jobs/7324870384) on CI.

**Comment by jorgeorpinel** [comment]
And several other PRs show the same issue.

**Comment by shcheklein** [comment]
@jorgeorpinel I've tried to rebase this PR made by @rogermparent and seems to pass the tests. Could you please rebase `guide/data-mgmt-flows` and this branch? 

> Several other PRs show the same issue.

Could you share which one of the PRs has this issue? I'll try to run some tests...

**Comment by jorgeorpinel** [comment]
Hmm actually it's only this one and https://github.com/iterative/dvc.org/pull/4119 (same [check fail](https://github.com/iterative/dvc.org/actions/runs/4219603359/jobs/7325127508)) ATM.

**Comment by jorgeorpinel** [comment]
I think they're both up to date with `main` but I'll try to rebase or merge (through the target branch in this case)... ⏳

**Comment by jorgeorpinel** [comment]
OK the problem only persists in this PR now. I may have found the problem in an accidental merge I did [some time ago](https://github.com/iterative/dvc.org/pull/4053#commits-pushed-9fc77a9) directly from `main` into this branch... There should not be 27 files changed here 😅

Let me see if I can still fix it. Else I'll re-do the PR. ⏳

File: content/docs/user-guide/data-management/track-sync-data.md [hunk 10]
```diff
@@ -0,0 +1,164 @@
+# Track and Sync Versioned Data & Models
+
+The fundamental workflow of most <abbr>DVC projects</abbr> includes the
+following **basic operations**. These can be performed directly (as we cover
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/track-sync-data.md:4]: This is the main file being contributed.

File: content/docs/user-guide/data-management/cloud-versioning.md [hunk 11]
```diff
@@ -30,19 +30,22 @@ benefits of content-addressable storage.
...
+However, this comes with the drawback of losing human-readable filenames without
+the use of the DVC CLI (`dvc get --show-url`) or API (`dvc.api.get_url()`).
 
 When using cloud versioning, DVC does not provide de-duplication, and certain
 remote storage performance optimizations will be unavailable.
 
+[content-addressable storage]:
+  /doc/user-guide/project-structure/internal-files#structure-of-the-cache-directory
+[data sync operations]:
+  /doc/user-guide/data-management/track-sync-data#synchronizing-data
```
Comment (jorgeorpinel) [inline content/docs/user-guide/data-management/cloud-versioning.md:47]: Most other changes just add links to the new page (mainly to the Sync section).

File: content/docs/user-guide/experiment-management/sharing-experiments.md [hunk 12]
```diff
@@ -4,7 +4,7 @@ In a regular Git workflow, <abbr>DVC repository</abbr> versions are typically
 synchronized among team members. And [DVC Experiments] are internally connected
 to this commit history, so you can similarly share them.
 
-## Basic workflow: store as peristent commits
+## Basic workflow: store as persistent commits
```
Comment (jorgeorpinel) [inline content/docs/user-guide/experiment-management/sharing-experiments.md:7]: This is an unrelated fix, oops. Found it when looking for places to link from (but didn't end up linking from this file).

**Review by jorgeorpinel — COMMENTED** [review]
Alright the PR looks good now. See the [pending questions above](https://github.com/iterative/dvc.org/pull/4053#issuecomment-1436232099) though. Summary of changes thus far:

**Comment by dberenbaum** [comment]
My take on this one: I don't see anywhere else in the data management guide that we have the basic workflow explained, so I think this one is useful for that. Happy to try and get it polished if you agree @shcheklein.

I would consider incorporating https://dvc.org/doc/user-guide/how-to/update-tracked-data and/or https://dvc.org/doc/user-guide/how-to/stop-tracking-data.

**Comment by shcheklein** [comment]
@dberenbaum sound good to me!

**Comment by dberenbaum** [comment]
@jorgeorpinel Do you want to finish this one, or would you rather I take it over?

**Comment by tapadipti** [comment]
@dberenbaum is this PR still relevant? If yes, maybe you could complete it [as you suggested](https://github.com/iterative/dvc.org/pull/4053#issuecomment-1465000861)?

**Comment by dberenbaum** [comment]
@tapadipti It is relevant but it doesn't seem like we are able to prioritize it right now, so I'll close.
